### PR TITLE
Token price processor - Observation [CCIP-2958]

### PR DESCRIPTION
**Pull request by asoliman92**
**TokenPrice Observation phase**

* Add GetFeeQuoterTokenUpdates in onchain_prices_reader.go
* Observe feedPrices, feeQuoterUpdates, fDestChain and current timestamp (UTC)

Note: We're observing Fchain, although we already observe it in merkleroot processor. This is because it's more straight forward if each processor takes care of what they need. For example the token price processor don't need all the Fs for all chains, just the destination fChain. Also reduces complexity of the PluginProcessor design as sending an extra parameter will require changing the current interface. If we see that it's affecting performance in the future then we can change the interface.

### CODE REVIEW

File: commit/tokenprice/processor.go [hunk 1]
```diff
@@ -47,6 +87,97 @@ func (w *Processor) ValidateObservation(
 	return nil
 }
 
+func (p *Processor) ObserveFDestChain() (*int, error) {
```
Comment (makramkd) [inline commit/tokenprice/processor.go:90]: This is a really wacky function signature, why return a pointer to an int?

File: commit/tokenprice/processor.go [hunk 2]
```diff
@@ -47,6 +87,97 @@ func (w *Processor) ValidateObservation(
...
+	tokensToQuery := maps.Keys(p.cfg.OffchainConfig.TokenInfo)
+	p.lggr.Infow("observing feed token prices")
+	tokenPrices, err := p.tokenPriceReader.GetTokenFeedPricesUSD(ctx, tokensToQuery)
+	if err != nil {
+		p.lggr.Errorw("call to GetTokenFeedPricesUSD failed", "err", err)
+		return []cciptypes.TokenPrice{}
+	}
+
+	// Continue if we couldn't fetch all prices
+	if len(tokenPrices) != len(tokensToQuery) {
```
Comment (makramkd) [inline commit/tokenprice/processor.go:133]: The comment and the line after don't line up. What does "continue" mean in this context?

File: commit/tokenprice/processor.go [hunk 3]
```diff
@@ -3,42 +3,82 @@ package tokenprice
...
 	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/libocr/commontypes"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
 
-	"github.com/smartcontractkit/chainlink-ccip/shared"
+	"golang.org/x/exp/maps"
 )
 
 type Processor struct {
```
Comment (makramkd) [inline commit/tokenprice/processor.go:23]: This type doesn't need to be exported, can be private. See https://github.com/smartcontractkit/chainlink-ccip/pull/108#discussion_r1749967753 for more info

File: commit/tokenprice/types.go [hunk 4]
```diff
@@ -14,10 +17,23 @@ type Outcome struct {
...
 
 type Observation struct {
-	// FeedTokenPrices for tokens from the feeds on the feed chain
-	FeedTokenPrices []cciptypes.TokenPrice `json:"feedTokenPrices"`
-	// PriceRegistryTokenUpdates for tokens from the PriceRegistry on the dest chain
-	PriceRegistryTokenUpdates []cciptypes.TokenPrice `json:"priceRegistryTokenPrices"`
-	// Observation time
-	Timestamp time.Time
+	FeedTokenPrices       []cciptypes.TokenPrice            `json:"feedTokenPrices"`
+	FeeQuoterTokenUpdates map[types.Account]NumericalUpdate `json:"FeeQuoterTokenUpdates"`
```
Comment (makramkd) [inline commit/tokenprice/types.go:21]:
```suggestion
	FeeQuoterTokenUpdates map[types.Account]NumericalUpdate `json:"feeQuoterTokenUpdates"`
```

File: commit/tokenprice/types.go [hunk 5]
```diff
@@ -14,10 +17,23 @@ type Outcome struct {
...
+	// ObserveFeedTokenPrices returns the latest token prices from the feed chain
+	ObserveFeedTokenPrices(ctx context.Context) []cciptypes.TokenPrice
+
+	// ObserveFeeQuoterTokenUpdates returns the latest token prices from the FeeQuoter on the dest chain
+	ObserveFeeQuoterTokenUpdates(ctx context.Context) map[types.Account]NumericalUpdate
+
+	ObserveFDestChain() (*int, error)
+}
+
+type NumericalUpdate struct {
```
Comment (makramkd) [inline commit/tokenprice/types.go:36]: name suggestion:

```suggestion
type TimestampedBig struct {
```

File: internal/reader/onchain_prices_reader.go [hunk 6]
```diff
@@ -4,40 +4,57 @@ import (
...
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
 	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
 
 	"golang.org/x/sync/errgroup"
 )
 
-type TokenPrices interface {
-	// GetTokenPricesUSD returns the prices of the provided tokens in USD.
+type PriceReader interface {
+	// GetTokenFeedPricesUSD returns the prices of the provided tokens in USD normalized to e18. 1USD=1e18USD.
```
Comment (makramkd) [inline internal/reader/onchain_prices_reader.go:23]: > 1USD=1e18USD

This doesn't really make much sense as an equation - maybe just copy/paste the comment in the FeeQuoter.sol contract?

File: internal/reader/onchain_prices_reader.go [hunk 7]
```diff
@@ -4,40 +4,57 @@ import (
...
+	GetFeeQuoterTokenUpdates(
+		ctx context.Context,
+		tokens []ocr2types.Account,
+	) (map[ocr2types.Account]NumericalUpdate, error)
+}
+
+// NumericalUpdate On-chain struct `TimestampedPackedUint224`:
+// nolint:lll
+// https://github.com/smartcontractkit/chainlink/blob/2d77ff4623d0a0032533c89f32fc251400382455/contracts/src/v0.8/ccip/libraries/Internal.sol#L43-L47
+type NumericalUpdate struct {
```
Comment (makramkd) [inline internal/reader/onchain_prices_reader.go:37]: This type is duplicated across different files/packages

Comment (asoliman92) [reply internal/reader/onchain_prices_reader.go:37]: Yes, it's going to be cyclic dependency if I have a new package for the shared types and use it in the internal/reader and commit as commit uses internal/reader.

File: commit/tokenprice/processor.go (hunk 1 above)
Comment (makramkd) [reply commit/tokenprice/processor.go:91]: Yeah, return 0. If an error is returned, the value should never be used as its undefined.

File: internal/reader/onchain_prices_reader.go (hunk 7 above)
Comment (makramkd) [reply internal/reader/onchain_prices_reader.go:37]: I don't think that's true. There is no cycle if you have the following organization:
* shared
* internal/reader imports shared
* commit imports internal/reader and shared
To get a cycle you need to have e.g commit import internal/reader which imports commit.

Comment (asoliman92) [reply internal/reader/onchain_prices_reader.go:37]: fixing.

File: commit/tokenprice/processor.go (hunk 3 above)
Comment (asoliman92) [reply commit/tokenprice/processor.go:25]: Not sure I get what you want here. I use it to do `maps.Keys(..)` inside the functions.

**Review by makramkd — APPROVED** [review]

**Review by makramkd — APPROVED** [review]

**Pull request merged**